c2rust-analyze/src/rewrite/shim.rs

```diff
@@ -0,0 +1,240 @@
+use crate::context::{FlagSet, GlobalAnalysisCtxt, GlobalAssignment};
+use crate::rewrite::expr::{self, CastBuilder};
+use crate::rewrite::Rewrite;
+use crate::type_desc::{self, TypeDesc};
+use crate::LTy;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::intravisit;
+use rustc_hir::{Expr, ExprKind, FnRetTy};
+use rustc_middle::hir::nested_filter;
+use rustc_middle::ty::{DefIdTree, TyCtxt, TypeckResults};
+use rustc_span::Span;
+use std::collections::HashSet;
+use std::iter;
+use std::mem;
+
+struct ShimCallVisitor<'a, 'tcx> {
+    gacx: &'a GlobalAnalysisCtxt<'tcx>,
+    gasn: &'a GlobalAssignment,
+    typeck_results: &'tcx TypeckResults<'tcx>,
+    rewrites: Vec<(Span, Rewrite)>,
+    mentioned_fns: HashSet<DefId>,
+}
+
+impl<'a, 'tcx> ShimCallVisitor<'a, 'tcx> {
+    fn handle_def_mention(&mut self, def_id: DefId, span: Span) {
+        let tcx = self.gacx.tcx;
+
+        // We only care about mentions of local functions, not including trait methods.
+        if !def_id.is_local() {
+            return;
+        }
+
+        match tcx.def_kind(def_id) {
+            DefKind::Fn => {}
+            DefKind::AssocFn => {
+                let parent_def_id = tcx.parent(def_id);
+                if tcx.def_kind(parent_def_id) == DefKind::Trait {
+                    return;
+                }
+                if tcx.impl_trait_ref(parent_def_id).is_some() {
+                    // Ignore calls to trait methods.
+                    return;
+                }
+            }
+            _ => return,
+        }
+
+        // Only functions whose signatures might change are relevant here.  Check that the function
+        // has at least one non-`FIXED` pointer in its signature.
+        let lsig = match self.gacx.fn_sigs.get(&def_id) {
+            Some(x) => x,
+            None => return,
+        };
+        let has_non_fixed_ptr = lsig
+            .inputs
+            .iter()
+            .cloned()
+            .chain(iter::once(lsig.output))
+            .flat_map(|lty| lty.iter())
+            .any(|lty| {
+                let ptr = lty.label;
+                !ptr.is_none() && !self.gasn.flags[ptr].contains(FlagSet::FIXED)
+            });
+        if !has_non_fixed_ptr {
```

Comment:
nit: I think `if has_all_fixed_ptrs` would be a bit clearer here, flipping the `.any` to `.all`.